ccmlib/cluster.py

```diff
@@ -106,36 +106,86 @@ def set_install_dir(self, install_dir=None, version=None, verbose=False):
 
     def actively_watch_logs_for_error(self, on_error_call, interval=1):
         """
-        Begins a timed daemon thread that scans logs for errors every interval seconds.
+        Begins a thread that repeatedly scans system.log for new errors, every interval seconds.
+        (The first pass covers the entire log contents written at that point,
+        subsequent scans cover newly appended log messages).
 
-        If an error is seen and the callback is executed, it will be called with an
-        OrderedDictionary mapping node name to a list of error lines.
+        Reports new errors, by calling the provided callback with an OrderedDictionary
+        mapping node name to a list of error lines.
+
+        Returns the thread itself, which should be .join()'ed to wrap up execution,
+        otherwise will run until the main thread exits.
         """
-        log_positions = defaultdict(int)
+        class LogWatchingThread(threading.Thread):
+            """
+            This class is embedded here for now, because it is used only from
+            within Cluster, and depends on cluster.nodelist().
+            """
+            def __init__(self, cluster):
+                super(LogWatchingThread, self).__init__()
+                self.cluster = cluster
+                self.daemon = True  # set so that thread will exit when main thread exits
+                self.req_stop_event = threading.Event()
+                self.done_event = threading.Event()
+                self.log_positions = defaultdict(int)
+
+            def scan(self):
+                errordata = OrderedDict()
 
-        def grep_and_mark_all_logs():
-            errordata = OrderedDict()
+                try:
+                    for node in self.cluster.nodelist():
+                        scan_from_mark = self.log_positions[node.name]
+                        next_time_scan_from_mark = node.mark_log()
+                        if (next_time_scan_from_mark == scan_from_mark):
+                            # log hasn't advanced, nothing to do for this node
+                            continue
+                        else:
+                            errors = node.grep_log_for_errors_from(seek_start=scan_from_mark)
```

Comment:
I am now confident that this logic only has a risk of overscanning, which is fine